iguana.corecontroller/src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java

```diff
@@ -0,0 +1,207 @@
+import org.aksw.iguana.cc.query.source.impl.FolderQuerySource;
+import org.aksw.iguana.commons.factory.TypedFactory;
+import org.apache.jena.rdf.model.Model;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This QueryHandler which is used by every worker who is extending the AbstractWorker.
+ * It initializes the QuerySource, QuerySelector, QuerySet and, if needed, PatternHandler.
+ * After the initialization, it provides the next query to the worker using the generated QuerySource
+ * and the order given by the QuerySelector.
+ *
+ * @author frensing
+ */
+public class QueryHandler {
+
+    protected final Logger LOGGER = LoggerFactory.getLogger(QueryHandler.class);
+
+    protected Map<String, Object> config;
+    protected Integer workerID;
+    protected String location;
+    protected int hashcode;
+
+    protected boolean caching;
+
+    protected QuerySelector querySelector;
+
+    protected QuerySet querySet;
+
+    protected LanguageProcessor langProcessor;
+
+    public QueryHandler(Map<String, Object> config, Integer workerID) {
+        this.config = config;
+        this.workerID = workerID;
+
+        this.location = (String) config.get("location");
+
+        if (config.containsKey("pattern")) {
+            initPattern();
+        } else {
+            initQuerySet();
+        }
+
+        initQuerySelector();
+        initLanguageProcessor();
+
+        this.hashcode = this.querySet.hashCode();
+    }
+
+    public void getNextQuery(StringBuilder queryStr, StringBuilder queryID) throws IOException {
+        int queryIndex = this.querySelector.getNextIndex();
+        queryStr.append(this.querySet.getQueryAtPos(queryIndex));
+        queryID.append(getQueryId(queryIndex));
+    }
+
+    public Model getTripleStats(String taskID) {
+        List<QueryWrapper> queries = new ArrayList<>(this.querySet.size());
+        for (int i = 0; i < this.querySet.size(); i++) {
+            try {
+                queries.add(new QueryWrapper(this.querySet.getQueryAtPos(i), getQueryId(i)));
+            } catch (Exception e) {
+                LOGGER.error("Could not parse query " + this.querySet.getName() + ":" + i, e);
+            }
+        }
+        return this.langProcessor.generateTripleStats(queries, "" + this.hashcode, taskID);
+    }
+
+    @Override
+    public int hashCode() {
+        return this.hashcode;
+    }
+
+    public int getQueryCount() {
+        return this.querySet.size();
+    }
+
+    public LanguageProcessor getLanguageProcessor() {
+        return this.langProcessor;
+    }
+
+    /**
+     * Will initialize the PatternHandler if a pattern config is given.
+     * The PatternHandler uses the original query source and generates a new query source
+     * which is used during the benchmark execution.
+     */
+    private void initPattern() {
```

Comment:
For this function it is a precondition that this.config.get("pattern") returns a result. That should be reflected in the docstring + in an assertion.